### PR TITLE
Fix overwriting text on paste

### LINKED ISSUE #4301: 4.15.0 - Paste overwrites existing text
## Reproduction steps

1. Copy this text line.
2. Open https://jsfiddle.net/laurin1/x45pjdef/
3. Paste copied line into the editor.
4. Press enter.
5. Paste the same line again.
6. Repeat 5-6 multiple times.

## Expected

Every time a new line is pasted.

## Actual

Pasted text replaces the line above.

## Other

The issue is reproducible with enter mode `CKEDITOR.ENTER_DIV` option and when the editor is initialized totally empty.

### PR DESCRIPTION
**Pull request by Comandeer**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4301](https://github.com/ckeditor/ckeditor4/issues/4301): Fixed: pasted content overwritten when pasted in initially empty editor with [`div` enter mode](https://ckeditor.com/docs/ckeditor4/latest/features/enterkey.html).
```

## What changes did you make?

I've removed the invocation of `getData()` and replaced it with direct checking of editable's HTML.

## Which issues does your PR resolve?

Closes #4301.

### CODE REVIEW

**Comment by jacekbogdanski** [comment]
Rebased onto the latest `master`.

**Review by jacekbogdanski — CHANGES_REQUESTED** [review]
Looks good, just some minor changes.

File: core/editable.js [hunk 1]
```diff
@@ -1647,7 +1647,12 @@
 		// guarantee it's result to be a valid DOM tree.
 		function insert( editable, type, data, range ) {
 			var editor = editable.editor,
-				dontFilter = false;
+				dontFilter = false,
+				html = editable.getHtml(),
+				// Instead of getData method, we directly check the HTML
+				// due to the fact that interal getData operates on latest snapshot,
+				// not the current content (#4301).
+				isEmptyEditable = html === '' || html.match( emptyParagraphRegexp );
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply core/editable.js:1655]: Actually `emptyParagraphRegexp` searches for empty `div`, `p` and several other elements. It's not a paragraph as an element, but just as a block of text.

[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply core/editable.js:1655]: Thanks for explaining that, but still we are missing unit and manual tests for that case. I just wanted to add them quickly during this review, but find out that I'm actually not able to create any easily reproducible test case for that scenario. Could you update your PR with these tests for empty container elements? If it's just defensive programming, it would be still good to have some spying tests instead.

**Review by jacekbogdanski — CHANGES_REQUESTED** [review]
I see that test `/tests/core/editable/manual/enterdiv` is failing on Chrome - but it seems like it's also failing on `master` branch. Not sure if the test started to failing or it has been poorly written. Maybe do you remember @Comandeer how this test should behave?

**Comment by Comandeer** [comment]
> I see that test `/tests/core/editable/manual/enterdiv` is failing on Chrome - but it seems like it's also failing on `master` branch. Not sure if the test started to failing or it has been poorly written.

In fact it was a regression introduced in #3380. Unit tests didn't check for it. I've added the missing test and fixed the current solution to include the case.

I've also added tests for empty paragraph regex.

File: tests/core/editable/enterdiv.js [hunk 2]
```diff
@@ -51,23 +86,47 @@ function testInsertHtml( htmlString ) {
...
+				editor.insertHtml( html );
+
+				assert.areSame( html, editor.getData() );
+			};
+
+			return CKEDITOR.tools.object.merge( tests, test );
+		}, {} );
+
+		return CKEDITOR.tools.object.merge( tests, tagTests );
+	}, {} );
```
Comment (jacekbogdanski) [inline tests/core/editable/enterdiv.js:127]: Maybe this test generation code would be a bit more readable if we would extract the cartesian product as a separate helper function? So we could just write `cartesianProduct( arr1, arr2 )` producing e.g. array-based tuple?

**Review by jacekbogdanski — CHANGES_REQUESTED** [review]
The test `/tests/core/editable/manual/enterdiv` is still failing on Firefox, IE (not all editors).

If we've introduced regression recently and it has been already published, we should create a separate ticket for that. Could you do it with some explanation of why the regression happened, as you already checked it out and have more knowledge on the topic?

**Comment by f1ames** [comment]
@Comandeer, @jacekbogdanski when you will be ready with this PR let's ask @ckeditor/qa-team (@LukaszGudel) to check it too :+1:

**Comment by Comandeer** [comment]
Ok, it seems that the issue is connected with the selection and it's not fixable in a sensible way: the selection is anchored to the text nodes inside `div` elements. It's not possible to  detect the intention of the user (to select only the text or everything), so I'm for merging the current fix as is.  
  
WDYT, @jacekbogdanski @f1ames?

**Comment by f1ames** [comment]
@Comandeer I'm not really in the context of this PR, could you sum up shortly what are the issues we have with the current solution? And what it fixes and what it breaks?

**Comment by Comandeer** [comment]
@f1ames, the main issue with the current fix is it being a partial one. It fixes only situation, when editor's content is selected with <kbd>Ctrl</kbd>+<kbd>A</kbd> or with Select All plugin. It doesn't work when the selection is made by hand.

**Comment by f1ames** [comment]
> @f1ames, the main issue with the current fix is it being a partial one. It fixes only situation, when editor's content is selected with Ctrl+A or with Select All plugin. It doesn't work when the selection is made by hand.

Well, it means it covers 2 of 3 ways one can select content so it will be a visible improvement. However, the initial issue reported (#4301) doesn't involve any ranged selection and it's about collapsed selection only :thinking: Also looking at the code changes in this PR it seems to not be connected with selection in any way - only the mechanism checking if editor has any data (is not empty) was changed which shouldn't be affected by the selection itself.

> It doesn't work when the selection is made by hand.

The [https://github.com/ckeditor/ckeditor4/pull/4356#issuecomment-750860386](https://github.com/ckeditor/ckeditor4/pull/4356#issuecomment-750860386) and the above made mea little confused how it relates to this fix TBH :thinking: @Comandeer could you elaborate a little more one this? What role the selection plays in this entire issue and how it effects the current fix?

**Comment by jacekbogdanski** [comment]
I did some wrap up about the whole issue with the current implementation, so we can finally proceed with this PR:

1. The first fix was for https://github.com/ckeditor/ckeditor4/issues/2751, **but**
    1. We introduced a regression for https://github.com/ckeditor/ckeditor4/issues/3379
    2. It didn't totally fix the issue. Some cases like selecting all content with the mouse (not ctrl+A) don't work correctly e.g. on IE11
2. We did fix for https://github.com/ckeditor/ckeditor4/issues/3379, **but**
    1. We introduced a regression for https://github.com/ckeditor/ckeditor4/issues/4301
3. We are here now, fixing https://github.com/ckeditor/ckeditor4/issues/4301, **but**
    1. We experienced all additional edge cases not fixed by partial https://github.com/ckeditor/ckeditor4/issues/2751 fix.
    
The best approach would be revisiting https://github.com/ckeditor/ckeditor4/issues/2751 and going with the proper fix in the first place. However, since the proper fix may not be trivial and the current issue looks like something which may occur much more often when using the editor, I would be for keeping the current fix and extracting the proper solution as a separate ticket. To compare issues:

* **Fix in #2751** - is partial and is not covering cases when whole editor selection is made by a mouse due to different selection anchors. **Results** in incorrectly wrapped HTML with redundant elements.
* **Current fix** - fixes the case when pasting the same text again into the empty editor results in losing pasted data.

So, it makes more sense to go with the current workaround (since it's almost finished) and start thinking about a more complicated, general fix for selection (replacing #2751 entirely), then postponing the fix for the current issue and working on a new fix #2751 right now.

I will go with another review round and if there are only minor issues left (like improving tests), we can go with this PR as the current solution.

File: tests/core/editable/enterdiv.js (hunk 2 above)
Comment (jacekbogdanski) [reply tests/core/editable/enterdiv.js:127]: After a second thought, it probably doesn't make much sense as the code is readable enough. Otherwise, we would have to avoid changing `tests` state directly which may complicate this test further.

**Comment by jacekbogdanski** [comment]
Looks like we are ready for some QA, @ckeditor/qa-team team could you help with testing this fix? We are focused on fixing https://github.com/ckeditor/ckeditor4/issues/4301 but as explained in https://github.com/ckeditor/ckeditor4/pull/4356#issuecomment-785789561 we still may have some issues with invalid div wrapping, but it should not be a part of this PR. I will extract a follow-up in a moment.

**Comment by LukaszGudel** [comment]
Other than the scenario from this issue (paste some text from outside) I've found three other ways of inserting content to the editor that are also affected by this bug. Additionally, it is possible to:
- drag and drop some text from outside the editor multiple times
- insert text using `editor.insertHtml('content')` multiple times
- type some text and then copy it and paste at the end multiple times

In all of the above cases the result is the same, only the last one copied/inserted text is showing.

On the ` t/4301` branch none of those cases produces this issue. I did not find any other issues with empty editor and div_enter.

**Review by jacekbogdanski — APPROVED** [review]
Thanks, @LukaszGudel! I see that failing CI is already reported:

https://github.com/ckeditor/ckeditor4/issues/3424
https://github.com/ckeditor/ckeditor4/issues/3868

Good job @Comandeer!
